### PR TITLE
[FLINK-27812] Support Dynamic Change of Watched Namespaces

### PR DESCRIPTION
**Pull request by morhidi**
This PR enables handling namespace changes via Dynamically Operator Configuration through the config option `kubernetes.operator.watched.namespaces`. Watched namespaces originally were defined using environment variables, thus namespaces changes required a POD restart always. With the current feature Helm upgrades won't trigger an automatic POD restart on namespace changes anymore. The feature is disabled by default, can be enabled by setting `kubernetes.operator.dynamic.namespaces.enabled=true` 

Suggestions for testing:

```
helm install flink-kubernetes-operator helm/flink-kubernetes-operator --set "watchNamespaces={default}"
helm upgrade flink-kubernetes-operator helm/flink-kubernetes-operator --set "watchNamespaces={default,flink}"
```
The helm upgrade will update the config map, but it takes some time until the new configs are propagated in the operator. Worth setting the `kubernetes.operator.dynamic.config.check.interval: 10 s` for local testing.

I have also moved the InformerManager into the webhook module since it is only used there, and opened a [FR](https://github.com/java-operator-sdk/java-operator-sdk/issues/1286) in JOSDK to get rid of it on long term.

I would be happy to get your feedback on this @wangyang0918 @Aitozi @gyfora @tweise

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 1]
```diff
@@ -187,4 +188,18 @@ public class KubernetesOperatorConfigOptions {
...
+                    .defaultValue(Constants.WATCH_ALL_NAMESPACES)
+                    .withDescription(
+                            "Comma separated list of namespaces the operator monitors for custom resources. Defaults to "
+                                    + Constants.WATCH_ALL_NAMESPACES);
+
+    public static final ConfigOption<Boolean> OPERATOR_DYNAMIC_NAMESPACES_ENABLED =
+            ConfigOptions.key("kubernetes.operator.dynamic.namespaces.enabled")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription("Enables dynamic change of namespaces. Defaults to false");
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:204]: nit:  `Enables dynamic change of watched/monitored namespaces`

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 2]
```diff
@@ -109,14 +117,19 @@ public Configuration getDefaultConfig() {
 
     @VisibleForTesting
     public void updateDefaultConfig(Configuration newConf) {
-        if (newConf.equals(defaultConfig)) {
+        if (defaultConfig != null
+                && newConf != null
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:121]: Do we need these nullchecks here? Why would they ever be null?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 3]
```diff
@@ -109,14 +117,19 @@ public Configuration getDefaultConfig() {
...
             LOG.info("Default configuration did not change, nothing to do...");
             return;
         }
 
         LOG.info("Updating default configuration to {}", newConf);
-        this.operatorConfiguration =
-                FlinkOperatorConfiguration.fromConfiguration(newConf, namespaces);
+        this.operatorConfiguration = FlinkOperatorConfiguration.fromConfiguration(newConf);
+
+        if (this.operatorConfiguration.getDynamicNamespacesEnabled()) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:130]: should we check if thhe watched namespaces actually changed?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/informer/InformerManager.java [hunk 4]
```diff
@@ -104,4 +100,24 @@ Map<String, SharedIndexInformer<CR>> createRunnableInformer(
             return informers;
         }
     }
+
+    public void setNameSpaces(Set<String> watchedNamespaces) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/informer/InformerManager.java:104]: nit: `setNameSpaces` -> `setNamespaces`

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 1 above)
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:204]: makes sense

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java (hunk 2 above)
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:121]: Yes, this is needed

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java (hunk 3 above)
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:130]: Sure, I'll add an extra check here

**Comment by morhidi** [comment]
Thank you for your review @gyfora I've addressed your comments.

**Pull request merged**